[docs] Migrate DML Documentation (#203)
moodle/devdocs: docs/apis/core/dml/index.md

## General concepts

### DB object

- The data manipulation API is exposed via public methods of the <tt>$DB</tt> object.
- Moodle core takes care of setting up the connection to the database according to values specified in the main config.php file.
- The $DB global object is an instance of the <tt>moodle_database</tt> class. It is instantiated automatically during the bootstrap setup, i.e. as a part of including the main config.php file.
- The DB object is available in the global scope right after including the config.php file:

```php title="example.php"
<?php
require(__DIR__ . '/config.php');

// You can access the database via the $DB method calls here.
```

- To make the DB object available in your local scope, such as within a function:

```php title="example.php"
<?php

function my_function_making_use_of_database() {
    global $DB;

    // You can access the database via the $DB method calls here.
}
```

### Table prefix

- Most Moodle installations use a prefix for all the database tables, such as <tt>mdl_</tt>. This prefix is NOT to be used in the code in the code itself.

> - All the $table parameters in the functions are meant to be the table name without prefixes:

Reviewer: ```suggestion
- All the `$table` parameters in the functions are meant to be the table name without prefixes:
```

```php
$user = $DB->get_record('user', [ => '1']('id'));
```

- In custom SQL queries, table names must be enclosed between curly braces. They will be then automatically converted to the real prefixed table name. There is no need to access <tt>$CFG->prefix</tt>

```php
$user = $DB->get_record_sql('SELECT COUNT(*) FROM {user} WHERE deleted = 1 OR suspended = 1;');
```

### Conditions

> - All the $conditions parameters in the functions are arrays of `fieldname  => fieldvalue` elements.

Reviewer: ```suggestion
- All the `$conditions` parameters in the functions are arrays of `fieldname  => fieldvalue` elements.
```

- They all must be fulfilled - that is the logical <tt>AND</tt> is used to populate the actual <tt>WHERE</tt> statement

```php
$user = $DB->get_record('user', 'firstname'  => 'Martin', 'lastname'  => 'Dougiamas');
```

### Placeholders

> - All the $params parameters in the functions are arrays of values used to fill placeholders in SQL statements.

Reviewer: ```suggestion
- All the `$params` parameters in the functions are arrays of values used to fill placeholders in SQL statements.
```

```php
public function insert_record(
    $table,
    $dataobject,
    $returnid = true,
    $bulk = false
);
```

### insert_records

Insert multiple records into the table as fast as possible. Records are inserted in the given order, but the operation is not atomic. Use transactions if necessary.

```php
public function insert_records(
    $table,
    $dataobjects
);
```

### insert_record_raw

For rare cases when you also need to specify the ID of the record to be inserted.

## Updating records

### update_record

Update a record in the table. The data object must have the property "id" set.

```php
public function update_record(
    $table,
    $dataobject,
    $bulk = false
);
```

## Executing a custom query

### execute

- If you need to perform a complex update using arbitrary SQL, you can use the low level "execute" method. Only use this when no specialised method exists.

```php
public function execute(
    $sql,
    array $params = null
);
```

:::danger

Do NOT use this to make changes in database structure, use the `database_manager` methods instead.

:::

## Using recordsets

> If the number of records to be retrieved from DB is high, the 'get_records_xxx() functions above are far from optimal, because they load all the records into the memory via the returned array. Under those circumstances, it is highly recommended to use these get_recordset_xxx() functions instead. They return an iterator to iterate over all the found records and save a lot of memory.

Reviewer: ```suggestion
If the number of records to be retrieved from DB is high, the `get_records_xxx()` functions above are far from optimal, because they load all the records into the memory via the returned array. Under those circumstances, it is highly recommended to use these `get_recordset_xxx()` functions instead. They return an iterator to iterate over all the found records and save a lot of memory.
```

It is **absolutely important** to not forget to close the returned recordset iterator after using it. This is to free up a lot of resources in the RDBMS.

> A general way to iterate over records using the get_recordset_xxx() functions:

Reviewer: ```suggestion
A general way to iterate over records using the `get_recordset_xxx()` functions:
```

    $field,
    array $values,
    $sort = *,
    $fields = '*',
    $limitfrom = *,
    $limitnum = ''
);
```

## Delegated transactions

- Please note some databases do not support transactions (such as the MyISAM MySQL database engine), however all server administrators are strongly encouraged to migrate to databases that support transactions (such as the InnoDB MySQL database engine).
- Previous versions supported only one level of transaction. Since Moodle 2.0, the DML layer emulates delegated transactions that allow nesting of transactions.
- Some subsystems (such as messaging) do not support transactions because it is not possible to rollback in external systems.
A transaction is started by:

```php
$transaction = $DB->start_delegated_transaction();
```

and finished by:

```php
$transaction->allow_commit();
```

Usually a transaction is rolled back when an exception is thrown:

```php
$transaction->rollback($ex);
```

which must be used very carefully because it might break compatibility with databases that do not support transactions. Transactions cannot be used as part of expected code flow; they can be used only as an emergency protection of data consistency.

> See more details in [DB layer 2.0 delegated transactions](https://docs.moodle.org/dev/DB_layer_2.0_delegated_transactions) or [MDL-20625](https://tracker.moodle.org/browse/MDL-20625).

Reviewer: ```suggestion
:::info More information:

See more details in [DB layer 2.0 delegated transactions](https://docs.moodle.org/dev/DB_layer_2.0_delegated_transactions) or [MDL-20625](https://tracker.moodle.org/browse/MDL-20625).

:::
```

Reviewer: I've also created https://github.com/moodle/devdocs/issues/217 with a proposal to migrate this page (adding the `good first issue` label too)

### Example

```php
global $DB;
try {
     $transaction = $DB->start_delegated_transaction();
     $DB->insert_record('foo', $object);
     $DB->insert_record('bar', $otherobject);

     // Assuming the both inserts work, we get to the following line.
     $transaction->allow_commit();

} catch(Exception $e) {
     $transaction->rollback($e);
}
```

## Cross-DB compatibility

> Moodle supports several SQL servers (MySQL, PostgreSQL, MS-SQL and Oracle). Each of them have some specific syntax in certain cases. In order to achieve cross-db compatibility of the code, following functions must be used to generate the fragments of the query valid for the actual SQL server.

Reviewer: ```suggestion
Moodle supports several SQL servers (MySQL, MariaDB, PostgreSQL, MS-SQL and Oracle). Each of them has some specific syntax in certain cases. In order to achieve cross-db compatibility of the code, the following functions must be used to generate the fragments of the query valid for the actual SQL server.
```

```php
public function sql_isnotempty(
    $tablename,
    $fieldname,
    $nullablefield,
    $textfield
);
```

### get_in_or_equal

Return the query fragment to check if a value is IN the given list of items (with a fallback to plain equal comparison if there is just one item)

```php
public function get_in_or_equal(
    $items,
    $type = SQL_PARAMS_QM,
    $prefix = 'param',
    $equal = true,
    $onemptyitems = false
);
```

Example:

```php
$statuses = ['todo', 'open', 'inprogress', 'intesting'];
[$insql, $inparams] = $DB->get_in_or_equal($statuses);
$sql = "SELECT * FROM {bugtracker_issues} WHERE status $insql";
$bugs = $DB->get_records_sql($sql, $inparams);
```

An example using named params:

```php
[$insql, $params] = $DB->get_in_or_equal($contexts, SQL_PARAMS_NAMED, 'ctx');
$contextsql = "AND rc.contextid {$insql}";
```

### sql_regex_supported

Does the current database driver support regex syntax when searching?

```php
public function sql_regex_supported()
```

### sql_regex

Return the query fragment to perform a regex search.

```php
public function sql_regex(
    $positivematch = true,
    $casesensitive = false
);
```

> Example: Searching for Page module instances containing links.
> 
> ```php

Reviewer: ```suggestion
```php title="Example: Searching for Page module instances containing links."
```
